lib/chef/knife/vsphere_vm_find.rb

```diff
@@ -0,0 +1,277 @@
+       
+option :matchip,
+       long: '--match-ip IP',
+       description: 'match ip'
+       
+option :matchos,
+       long: '--match-os OS',
+       description: 'match os'
+
+
+option :matchname,
+       long: '--match-name VMNAME',
+       description: 'match name'
+
+
+option :hostname,
+       long: '--hostname',
+       description: 'show hostname'
+
+option :os,
+       long: '--os',
+       description: 'show os details'
+
+option :alarms,
+       long: '--alarms',
+       description: 'show alarm status'
+
+option :tools,
+       long: '--tools',
+       description: 'show tools status'
+
+option :matchtools,
+       long: '--match-tools TOOLSSTATE',
+       description: 'match tools state'
+
+option :full_path,
+       long: '--full-path',
+       description: 'Show full path'
+
+
+
+$stdout.sync = true   #smoother output from print
+
+
+  def traverse_folders_for_pool(folder, poolname)
+    #children = folder.children.find_all
+    children = find_all_in_folder(folder, RbVmomi::VIM::ManagedObject)
+    children.each do |child|
+      if child.class == RbVmomi::VIM::ClusterComputeResource || child.class == RbVmomi::VIM::ComputeResource || child.class == RbVmomi::VIM::ResourcePool 
+        if child.name == poolname then return child 
+         else if child.class == RbVmomi::VIM::Folder || child.class == RbVmomi::VIM::ComputeResource || child.class == RbVmomi::VIM::ClusterComputeResource || child.class == RbVmomi::VIM::ResourcePool
+        pool = traverse_folders_for_pool(child, poolname)
+        end
+        end
+        if pool then return pool end
+      end
+    end
+    return false
+  end
+  
+  def find_pool_folder(folderName)
+    dc = get_datacenter
+    baseEntity = dc.hostFolder
+    entityArray = folderName.split('/')
+    entityArray.each do |entityArrItem|
+      if entityArrItem != ''
+        baseEntity = baseEntity.childEntity.grep(RbVmomi::VIM::ManagedObject).find { |f| f.name == entityArrItem } or
+            abort "no such folder #{folderName} while looking for #{entityArrItem}"
+      end
+    end
+    baseEntity
+  end
+
+  def run
+    poolname = config[:pool]
+    if poolname.nil?
+      show_usage
+      fatal_exit("You must specify a resource pool or cluster name (see knife vsphere pool list)")
+    end
+
+
+    vim = vim_connection
+    dc = datacenter
+    folder = dc.hostFolder
+
+pool = traverse_folders_for_pool(folder, poolname) or abort "Pool #{poolname} not found"
+
+
+if pool.class == RbVmomi::VIM::ResourcePool then
+  vm=pool.vm
+else vm=pool.resourcePool.vm
+end
+
+ unless vm.nil?
+      vm.each do |vmc|
+
+      state = case vmc.runtime.powerState
+                when PsOn
+                  ui.color("on", :green)
+                when PsOff
+                  ui.color("off", :red)
+                when PsSuspended
+                  ui.color("suspended", :yellow)
+              end
+
+
+if get_config(:matchname)
+  if vmc.name.include? config[:matchname]
+  else next
+  end
+end
```

Comment:
I think this, and the ones that follow, could be a bit more clear with `unless`, e.g.
```
if get_config(:matchname)
  next unless vmc.name.include? config[:matchname]
end
```